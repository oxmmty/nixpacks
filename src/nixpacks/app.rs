use std::path::Path;
use std::{env, fs, path::PathBuf};

use anyhow::{Context, Result};
use glob::glob;
use regex::Regex;
use serde::de::DeserializeOwned;

#[derive(Debug, Clone)]
pub struct App {
    pub source: PathBuf,
    pub paths: Vec<PathBuf>,
}

impl App {
    pub fn new(path: &str) -> Result<App> {
        let current_dir = env::current_dir()?;
        let source = current_dir
            .join(path)
            .canonicalize()
            .context("Failed to read app source directory")?;

        let dir = fs::read_dir(source.clone()).context("Failed to read app source directory")?;
        let paths: Vec<PathBuf> = dir.map(|path| path.unwrap().path()).collect();

        Ok(App { source, paths })
    }

    pub fn includes_file(&self, name: &str) -> bool {
        fs::canonicalize(self.source.join(name)).is_ok()
    }

    pub fn find_files(&self, pattern: &str) -> Result<Vec<String>> {
        let full_pattern = self.source.join(pattern);

        let pattern_str = match full_pattern.to_str() {
            Some(s) => s,
            None => return Ok(Vec::new()),
        };

        let relative_paths = glob(pattern_str)?
            .filter_map(|p| p.ok()) // Remove bad ones
            .filter_map(|p| p.to_str().map(|p| p.to_string())) // convert to string
            .collect();

        Ok(relative_paths)
    }

    pub fn read_file(&self, name: &str) -> Result<String> {
        let name = self.source.join(name);
        let contents = fs::read_to_string(name)?;
        Ok(contents)
    }

    pub fn find_match(&self, re: &Regex, pattern: &str) -> Result<bool> {
        let paths = match self.find_files(pattern) {
            Ok(v) => v,
            Err(_e) => return Ok(false),
        };

        for path in paths {
            let path_buf = fs::canonicalize(path)?;

            if let Some(p) = path_buf.to_str() {
                let f = self.read_file(p)?;
                let matches = re.find(f.as_str());
                if matches.is_some() {
                    return Ok(true);
                }
            }
        }

        Ok(false)
    }

    pub fn read_json<T>(&self, name: &str) -> Result<T>
    where
        T: DeserializeOwned,
    {
        let contents = self.read_file(name)?;
        let value: T = serde_json::from_str(contents.as_str())?;
        Ok(value)
    }

    pub fn read_toml<T>(&self, name: &str) -> Result<T>
    where
        T: DeserializeOwned,
    {
        let contents = self.read_file(name)?;
        let toml_file = toml::from_str(contents.as_str())?;
        Ok(toml_file)
    }

    pub fn strip_source_path(&self, abs: &str) -> Result<String> {
        let source_str = match self.source.to_str() {
            Some(s) => s,
            None => return Err(anyhow::Error::msg("Failed to parse source path")),
        };
        let abs_path = Path::new(abs);

        // Strip source path from absolute path
        let stripped = match abs_path.strip_prefix(source_str) {
            Ok(p) => p,
            Err(_e) => abs_path,
        };

        // Convert path to string
        return match stripped.to_str() {
            Some(v) => Ok(v.to_string()),
            None => Err(anyhow::Error::msg("Failed to convert path to string")),
        };
    }
}

#[cfg(test)]
mod tests {
    use std::collections::HashMap;

    use super::*;
    use serde::{Deserialize, Serialize};
    use serde_json::{Map, Value};

    #[derive(Serialize, Deserialize)]
    struct TestPackageJson {
        name: String,
        scripts: HashMap<String, String>,
    }

    #[test]
    fn test_creates_app() -> Result<()> {
        let app = App::new("./examples/npm")?;
        assert_eq!(app.paths.len(), 5);
        Ok(())
    }

    #[test]
    fn test_read_file() -> Result<()> {
        let app = App::new("./examples/npm")?;
        assert_eq!(
            app.read_file("index.ts")?.trim_end(),
            "console.log(\"Hello from NPM\");"
        );
        Ok(())
    }

    #[test]
    fn test_read_json_file() -> Result<()> {
        let app = App::new("./examples/npm")?;
        let value: Map<String, Value> = app.read_json("package.json")?;
        assert!(value.get("name").is_some());
        assert_eq!(value.get("name").unwrap(), "npm");
        Ok(())
    }

    #[test]
    fn test_read_structured_json_file() -> Result<()> {
        let app = App::new("./examples/npm")?;
        let value: TestPackageJson = app.read_json("package.json")?;
        assert_eq!(value.name, "npm");
        assert_eq!(value.scripts.get("build").unwrap(), "tsc -p tsconfig.json");
        Ok(())
    }

    #[test]
    fn test_read_toml_file() -> Result<()> {
        let app = App::new("./examples/rust-rocket")?;
        let toml_file: toml::Value = app.read_toml("Cargo.toml")?;
        assert!(toml_file.get("package").is_some());
        assert_eq!(
            toml_file
                .get("package")
                .unwrap()
                .get("name")
                .unwrap()
                .as_str()
                .unwrap(),
            "rocket"
        );
        Ok(())
    }

    #[test]
    fn test_find_files() -> Result<()> {
        let app = App::new("./examples/monorepo")?;
        let m = app.find_files("**/*.tsx").unwrap();
        assert_eq!(
            m,
            vec![
                "packages/client/pages/_app.tsx",
                "packages/client/pages/index.tsx"
            ]
        );
        Ok(())
    }

    #[test]
    fn test_find_match() -> Result<()> {
        let app = App::new("./examples/monorepo")?;
        let re = Regex::new(r"className")?;
        let m = app.find_match(&re, "**/*.tsx").unwrap();
        assert!(m);
        Ok(())
    }

    #[test]
    fn test_strip_source_path() -> Result<()> {
        let app = App::new("./examples/npm")?;
        let path_to_strip = app.source.join("foo/bar.txt");
        assert_eq!(
            app.strip_source_path(path_to_strip.to_str().unwrap())
                .unwrap(),
            "foo/bar.txt"
        );
        Ok(())
    }

    #[test]
    fn test_strip_source_path_no_source_prefix() -> Result<()> {
        let app = App::new("./examples/npm")?;
        assert_eq!(app.strip_source_path("no/prefix.txt")?, "no/prefix.txt");
        Ok(())
    }
}
