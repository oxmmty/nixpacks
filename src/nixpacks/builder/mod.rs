use super::{app::App, plan::BuildPlan};
use anyhow::Result;

pub mod docker;

pub trait Builder {
    fn create_image(&self, app: &App, plan: &BuildPlan) -> Result<()>;
}
