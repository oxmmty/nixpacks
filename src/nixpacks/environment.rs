use std::collections::HashMap;

pub type EnvironmentVariables = HashMap<String, String>;

#[derive(Default, Debug)]
pub struct Environment {
    variables: EnvironmentVariables,
}

impl Environment {
    pub fn new(variables: EnvironmentVariables) -> Environment {
        Environment { variables }
    }

    pub fn get_variable(&self, name: &str) -> Option<&String> {
        self.variables.get(name)
    }

    pub fn set_variable(&mut self, name: String, value: String) {
        self.variables.insert(name, value);
    }

    pub fn get_variable_names(&self) -> Vec<String> {
        self.variables.keys().cloned().collect()
    }

    pub fn clone_variables(env: &Environment) -> EnvironmentVariables {
        env.variables.clone()
    }
}

#[cfg(test)]
mod tests {
    use super::Environment;

    #[test]
    fn set_and_get_variables() {
        let mut environment = Environment::default();
        assert!(environment.get_variable("hello").is_none());
        environment.set_variable("hello".to_string(), "world".to_string());
        assert_eq!(
            environment.get_variable("hello"),
            Some(&"world".to_string())
        );
    }
}
