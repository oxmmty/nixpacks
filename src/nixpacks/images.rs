pub static DEFAULT_BASE_IMAGE: &str = "ghcr.io/railwayapp/nixpacks:debian-1652414952";

pub static DEBIAN_SLIM_IMAGE: &str = "debian:bullseye-slim";
