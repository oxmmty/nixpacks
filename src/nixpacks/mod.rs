use anyhow::{bail, Context, Ok, Result};
use indoc::formatdoc;
use std::{
    fs::{self, File},
    io::Write,
    path::{Path, PathBuf},
    process::Command,
};
use tempdir::TempDir;
use uuid::Uuid;
pub mod app;
pub mod logger;
pub mod plan;

use crate::providers::{Pkg, Provider};

use self::{app::App, logger::Logger, plan::BuildPlan};

static NIX_PACKS_VERSION: &str = "0.0.1";

// https://status.nixos.org/
static NIXPKGS_ARCHIVE: &str = "30d3d79b7d3607d56546dd2a6b49e156ba0ec634";

#[derive(Debug)]
pub struct AppBuilderOptions {
    pub custom_build_cmd: Option<String>,
    pub custom_start_cmd: Option<String>,
    pub custom_pkgs: Vec<Pkg>,
    pub pin_pkgs: bool,
}

impl AppBuilderOptions {
    pub fn empty() -> AppBuilderOptions {
        AppBuilderOptions {
            custom_build_cmd: None,
            custom_start_cmd: None,
            custom_pkgs: Vec::new(),
            pin_pkgs: false,
        }
    }
}

pub struct AppBuilder<'a> {
    name: Option<String>,
    app: &'a App,
    logger: &'a Logger,
    options: &'a AppBuilderOptions,
    provider: Option<&'a dyn Provider>,
}

impl<'a> AppBuilder<'a> {
    pub fn new(
        name: Option<String>,
        app: &'a App,
        logger: &'a Logger,
        options: &'a AppBuilderOptions,
    ) -> Result<AppBuilder<'a>> {
        Ok(AppBuilder {
            name,
            app,
            logger,
            options,
            provider: None,
        })
    }

    pub fn plan(&mut self, providers: Vec<&'a dyn Provider>) -> Result<BuildPlan> {
        // Load options from the best matching provider
        self.detect(providers).context("Detecting provider")?;

        let pkgs = self.get_pkgs().context("Getting packages")?;
        let install_cmd = self
            .get_install_cmd()
            .context("Generating install command")?;
        let build_cmd = self.get_build_cmd().context("Generating build command")?;
        let start_cmd = self.get_start_cmd().context("Generating start command")?;

        let plan = BuildPlan {
            version: NIX_PACKS_VERSION.to_string(),
            nixpkgs_archive: if self.options.pin_pkgs {
                Some(NIXPKGS_ARCHIVE.to_string())
            } else {
                None
            },
            pkgs,
            install_cmd,
            start_cmd,
            build_cmd,
        };

        Ok(plan)
    }

    pub fn build(&mut self, providers: Vec<&'a dyn Provider>) -> Result<()> {
        self.logger.log_section("Building");
        let plan = self.plan(providers).context("Creating build plan")?;
        self.logger.log_step("Generated new build plan");

        self.do_build(&plan)
    }

    pub fn build_from_plan(&mut self, plan: &BuildPlan) -> Result<()> {
        self.logger.log_section("Building");
        self.logger.log_step("Building from existing plan");

        self.do_build(plan)
    }

    pub fn do_build(&mut self, plan: &BuildPlan) -> Result<()> {
        let id = Uuid::new_v4();
        let tmp_dir = TempDir::new("nixpacks").context("Creating a temp directory")?;

        self.logger.log_step("Copying source to tmp dir");

        let source = self.app.source.as_path().to_str().unwrap();
        let mut copy_cmd = Command::new("cp")
            .arg("-a")
            .arg(format!("{}/.", source))
            .arg(tmp_dir.path())
            .spawn()?;
        let copy_result = copy_cmd.wait().context("Copying app source to tmp dir")?;
        if !copy_result.success() {
            bail!("Copy failed")
        }

        self.logger.log_step("Writing build plan");
        AppBuilder::write_build_plan(plan, tmp_dir.path()).context("Writing build plan")?;

        self.logger.log_step("Building image");

        let name = self.name.clone().unwrap_or_else(|| id.to_string());

        let mut docker_build_cmd = Command::new("docker")
            .arg("build")
            .arg(tmp_dir.path())
            .arg("-t")
            .arg(name.clone())
            .spawn()?;

        let build_result = docker_build_cmd.wait().context("Building image")?;

        if !build_result.success() {
            bail!("Docker build failed")
        }

        self.logger.log_section("Successfully Built!");

        println!("\nRun:");
        println!("  docker run {}", name);

        Ok(())
    }

    fn get_pkgs(&self) -> Result<Vec<Pkg>> {
        let pkgs: Vec<Pkg> = match self.provider {
            Some(provider) => {
                let mut provider_pkgs = provider.pkgs(self.app);
                let mut pkgs = self.options.custom_pkgs.clone();
                pkgs.append(&mut provider_pkgs);
                pkgs
            }
            None => self.options.custom_pkgs.clone(),
        };

        Ok(pkgs)
    }

    fn get_install_cmd(&self) -> Result<Option<String>> {
        let install_cmd = match self.provider {
            Some(provider) => provider.install_cmd(self.app)?,
            None => None,
        };

        Ok(install_cmd)
    }

    fn get_build_cmd(&self) -> Result<Option<String>> {
        let suggested_build_cmd = match self.provider {
            Some(provider) => provider.suggested_build_cmd(self.app)?,
            None => None,
        };

        let build_cmd = self
            .options
            .custom_build_cmd
            .clone()
            .or(suggested_build_cmd);

        Ok(build_cmd)
    }

    fn get_start_cmd(&self) -> Result<Option<String>> {
        let procfile_cmd = self.parse_procfile()?;

        let suggested_start_cmd = match self.provider {
            Some(provider) => provider.suggested_start_command(self.app)?,
            None => None,
        };

        let start_cmd = self
            .options
            .custom_start_cmd
            .clone()
            .or(procfile_cmd)
            .or(suggested_start_cmd);

        Ok(start_cmd)
    }

    fn detect(&mut self, providers: Vec<&'a dyn Provider>) -> Result<()> {
        for provider in providers {
            let matches = provider.detect(self.app)?;
            if matches {
                self.provider = Some(provider);
                break;
            }
        }

        Ok(())
    }

    fn parse_procfile(&self) -> Result<Option<String>> {
        if self.app.includes_file("Procfile") {
            let contents = self.app.read_file("Procfile")?;

            // Better error handling
            if contents.starts_with("web: ") {
                return Ok(Some(contents.replace("web: ", "").trim().to_string()));
            }

            Ok(None)
        } else {
            Ok(None)
        }
    }

    pub fn write_build_plan(plan: &BuildPlan, dest: &Path) -> Result<()> {
        let nix_expression = AppBuilder::gen_nix(plan).context("Generating Nix expression")?;
        let dockerfile = AppBuilder::gen_dockerfile(plan).context("Generating Dockerfile")?;

        let nix_path = PathBuf::from(dest).join(PathBuf::from("environment.nix"));
        let mut nix_file = File::create(nix_path).context("Creating Nix environment file")?;
        nix_file
            .write_all(nix_expression.as_bytes())
            .context("Unable to write Nix expression")?;

        let dockerfile_path = PathBuf::from(dest).join(PathBuf::from("Dockerfile"));
        File::create(dockerfile_path.clone()).context("Creating Dockerfile file")?;
        fs::write(dockerfile_path, dockerfile).context("Writing Dockerfile")?;

        Ok(())
    }

    pub fn gen_nix(plan: &BuildPlan) -> Result<String> {
        let nixpkgs = plan
            .pkgs
            .iter()
            .map(|p| p.name.clone())
            .collect::<Vec<String>>()
            .join(" ");

        let nix_archive = plan.nixpkgs_archive.clone();
        let pkg_import = match nix_archive {
            Some(archive) => format!(
                "with import (fetchTarball \"https://github.com/NixOS/nixpkgs/archive/{}.tar.gz\")",
                archive
            ),
            None => "with import <nixpkgs>".to_string(),
        };

        let nix_expression = formatdoc! {"
           {pkg_import} {{ }}; [ {pkgs} ]
        ",
        pkg_import=pkg_import,
        pkgs=nixpkgs};

        Ok(nix_expression)
    }

    pub fn gen_dockerfile(plan: &BuildPlan) -> Result<String> {
        let install_cmd = plan
            .install_cmd
            .as_ref()
            .map(|cmd| format!("RUN {}", cmd))
            .unwrap_or_else(|| "".to_string());
        let build_cmd = plan
            .build_cmd
            .as_ref()
            .map(|cmd| format!("RUN {}", cmd))
            .unwrap_or_else(|| "".to_string());
        let start_cmd = plan
            .start_cmd
            .as_ref()
            .map(|cmd| format!("CMD {}", cmd))
            .unwrap_or_else(|| "".to_string());

        let dockerfile = formatdoc! {"
          FROM nixos/nix

          RUN nix-channel --update

          COPY . /app
          WORKDIR /app

          # Load Nix environment
          RUN nix-env -if environment.nix

          # Install
          {install_cmd}

          # Build
          {build_cmd}

          # Start
          {start_cmd}
        ",
        install_cmd=install_cmd,
        build_cmd=build_cmd,
        start_cmd=start_cmd};

        Ok(dockerfile)
    }
}
