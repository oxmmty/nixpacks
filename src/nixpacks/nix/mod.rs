use super::plan::BuildPlan;
use indoc::formatdoc;

pub mod pkg;

pub fn create_nix_expression(plan: &BuildPlan) -> String {
    let setup_phase = plan.setup.clone().unwrap_or_default();

    let nixpkgs = setup_phase
        .pkgs
        .iter()
        .map(pkg::Pkg::to_nix_string)
        .collect::<Vec<String>>()
        .join(" ");

    let libraries = setup_phase.libraries.unwrap_or_default().join(" ");

    let nix_archive = setup_phase.archive.clone();
    let pkg_import = match nix_archive {
        Some(archive) => format!(
            "import (fetchTarball \"https://github.com/NixOS/nixpkgs/archive/{}.tar.gz\")",
            archive
        ),
        None => "import <nixpkgs>".to_string(),
    };

    let mut overlays: Vec<String> = Vec::new();
    for pkg in &setup_phase.pkgs {
        if let Some(overlay) = &pkg.overlay {
            overlays.push(overlay.to_string());
        }
    }

    let overlays_string = overlays
        .iter()
        .map(|url| format!("(import (builtins.fetchTarball \"{}\"))", url))
        .collect::<Vec<String>>()
        .join("\n");

    // If the openssl library is added, set the OPENSSL_DIR and OPENSSL_LIB_DIR environment variables
    // In the future, we will probably want a generic way for providers to set variables based off Nix package locations
    let openssl_dirs = if libraries.contains("openssl") {
        formatdoc! {"
          export OPENSSL_DIR=\"${{openssl.dev}}\"
          export OPENSSL_LIB_DIR=\"${{openssl.out}}/lib\"
        "}
    } else {
        String::new()
    };

    let nix_expression = formatdoc! {"
            {{ }}:

            let pkgs = {} {{ overlays = [ {} ]; }};
            in with pkgs;
              let
                APPEND_LIBRARY_PATH = \"${{lib.makeLibraryPath [ {} ] }}\";
                myLibraries = writeText \"libraries\" ''
                  export LD_LIBRARY_PATH=\"${{APPEND_LIBRARY_PATH}}:$LD_LIBRARY_PATH\"
                  {}
                '';
              in
                buildEnv {{
                  name = \"env\";
                  paths = [
                    (runCommand \"libraries\" {{ }} ''
                      mkdir -p $out/etc/profile.d
                      cp ${{myLibraries}} $out/etc/profile.d/libraries.sh
                    '')
                    {}
                  ];
                }}
        ",
        pkg_import,
        overlays_string,
        libraries,
        openssl_dirs,
        nixpkgs
    };

    nix_expression
}
