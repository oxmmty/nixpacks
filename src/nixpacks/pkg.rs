use std::collections::HashMap;

use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, PartialEq, Clone, Debug)]
pub struct Pkg {
    pub name: String,
    pub overrides: HashMap<String, String>,
}

impl Pkg {
    pub fn new(name: &str) -> Pkg {
        Pkg {
            name: name.to_string(),
            overrides: HashMap::default(),
        }
    }

    pub fn to_nix_string(&self) -> String {
        if self.overrides.is_empty() {
            self.name.clone()
        } else {
            let override_string = self
                .overrides
                .iter()
                .map(|(name, value)| format!("{} = {};", name, value))
                .collect::<Vec<_>>()
                .join(" ");
            format!("({}.override {{ {} }})", self.name, override_string)
        }
    }

    pub fn set_override(mut self, name: &str, pkg: &str) -> Self {
        self.overrides.insert(name.to_string(), pkg.to_string());
        self
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_simple_pkg_to_string() {
        assert_eq!(Pkg::new("cowsay").to_nix_string(), "cowsay".to_string());
    }

    #[test]
    fn test_pkg_single_override_to_string() {
        assert_eq!(
            Pkg::new("cowsay")
                .set_override("hello", "hello_1.1")
                .to_nix_string(),
            "(cowsay.override { hello = hello_1.1; })".to_string()
        );
    }
}
