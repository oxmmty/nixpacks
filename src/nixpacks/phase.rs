use serde::{Deserialize, Serialize};

use super::nix::Pkg;

#[serde_with::skip_serializing_none]
#[derive(Serialize, Deserialize, Default, Clone, Debug)]
pub struct SetupPhase {
    pub pkgs: Vec<Pkg>,
    pub archive: Option<String>,

    #[serde(rename = "onlyIncludeFiles")]
    pub only_include_files: Option<Vec<String>>,
}

impl SetupPhase {
    pub fn new(pkgs: Vec<Pkg>) -> Self {
        Self {
            pkgs,
            archive: None,
            only_include_files: None,
        }
    }

    pub fn add_file_dependency(&mut self, file: String) {
        if let Some(mut files) = self.only_include_files.clone() {
            files.push(file);
            self.only_include_files = Some(files);
        } else {
            self.only_include_files = Some(vec![file]);
        }
    }

    pub fn add_pkgs(&mut self, new_pkgs: &mut Vec<Pkg>) {
        self.pkgs.append(new_pkgs);
    }

    pub fn set_archive(&mut self, archive: String) {
        self.archive = Some(archive);
    }
}

#[serde_with::skip_serializing_none]
#[derive(Serialize, Deserialize, Default, Clone, Debug)]
pub struct InstallPhase {
    pub cmd: Option<String>,

    #[serde(rename = "onlyIncludeFiles")]
    pub only_include_files: Option<Vec<String>>,
}

impl InstallPhase {
    pub fn new(cmd: String) -> Self {
        Self {
            cmd: Some(cmd),
            only_include_files: None,
        }
    }

    pub fn add_file_dependency(&mut self, file: String) {
        if let Some(mut files) = self.only_include_files.clone() {
            files.push(file);
            self.only_include_files = Some(files);
        } else {
            self.only_include_files = Some(vec![file]);
        }
    }
}

#[serde_with::skip_serializing_none]
#[derive(Serialize, Deserialize, Default, Clone, Debug)]
pub struct BuildPhase {
    pub cmd: Option<String>,

    #[serde(rename = "onlyIncludeFiles")]
    pub only_include_files: Option<Vec<String>>,
}

impl BuildPhase {
    pub fn new(cmd: String) -> Self {
        Self {
            cmd: Some(cmd),
            only_include_files: None,
        }
    }

    pub fn add_file_dependency(&mut self, file: String) {
        if let Some(mut files) = self.only_include_files.clone() {
            files.push(file);
            self.only_include_files = Some(files);
        } else {
            self.only_include_files = Some(vec![file]);
        }
    }
}

#[serde_with::skip_serializing_none]
#[derive(Serialize, Deserialize, Default, Clone, Debug)]
pub struct StartPhase {
    pub cmd: Option<String>,
}

impl StartPhase {
    pub fn new(cmd: String) -> Self {
        Self { cmd: Some(cmd) }
    }
}
