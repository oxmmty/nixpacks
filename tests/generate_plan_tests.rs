use nixpacks::{generate_build_plan, nixpacks::plan::generator::GeneratePlanOptions};
use std::env::consts::ARCH;

test_helper::generate_plan_tests!();

#[test]
fn test_custom_rust_version() {
    let plan = simple_gen_plan("./examples/rust-custom-version");
    assert_eq!(
        plan.build.unwrap().cmds,
        Some(vec![
            format!("cargo build --release --target {}-unknown-linux-musl", ARCH),
            format!(
                "cp target/{}-unknown-linux-musl/release/rust-custom-version rust-custom-version",
                ARCH
            )
        ])
    );
    assert_eq!(
        plan.setup
            .unwrap()
            .pkgs
            .iter()
            .filter(|p| p.name.contains("1.56.0"))
            .count(),
        1
    );
}

#[test]
fn test_rust_rocket() {
    let plan = simple_gen_plan("./examples/rust-rocket");
    assert_eq!(
        plan.build.unwrap().cmds,
        Some(vec![
            format!("cargo build --release --target {}-unknown-linux-musl", ARCH),
            format!(
                "cp target/{}-unknown-linux-musl/release/rocket rocket",
                ARCH
            )
        ])
    );
    assert!(plan.start.clone().unwrap().cmd.is_some());
    assert_eq!(
        plan.start.clone().unwrap().cmd.unwrap(),
        "./rocket".to_string()
    );
    assert!(plan.start.unwrap().run_image.is_some());
}

#[test]
fn test_rust_rocket_no_musl() {
    let plan = generate_build_plan(
        "./examples/rust-rocket",
        vec!["NIXPACKS_NO_MUSL=1"],
        &GeneratePlanOptions::default(),
    )
    .unwrap();
    assert_plan_snapshot!(plan);
}

#[test]
fn test_rust_cargo_workspaces() {
    let plan = simple_gen_plan("./examples/rust-cargo-workspaces");

    assert_eq!(
        plan.build.unwrap().cmds.unwrap()[0],
        format!(
            "cargo build --release --package binary --target {}-unknown-linux-musl",
            ARCH
        )
    );
}

#[test]
fn test_haskell_stack() {
    let plan = simple_gen_plan("./examples/haskell-stack");
    assert_eq!(
        plan.install.unwrap().cmds,
        Some(vec!["stack setup".to_string()])
    );
    assert_eq!(
        plan.build.unwrap().cmds,
        Some(vec!["stack install".to_string()])
    );
    assert_eq!(
        plan.start.unwrap().cmd,
        Some("/root/.local/bin/haskell-stack-exe".to_string())
    );
}

#[cfg(any(target_arch = "aarch64", target_arch = "x86_64", target_arch = "i386"))]
#[test]
fn test_zig_gyro() {
    let plan = simple_gen_plan("./examples/zig-gyro");

    assert_eq!(
        plan.build.unwrap().cmds,
        Some(vec!["zig build -Drelease-safe=true".to_string()])
    );
    assert_eq!(
        plan.start.unwrap().cmd,
        Some("./zig-out/bin/zig-gyro".to_string())
    );
    assert!(plan
        .install
        .unwrap()
        .cmds
        .unwrap()
        .get(0)
        .unwrap()
        .contains("mkdir /gyro"));
}
