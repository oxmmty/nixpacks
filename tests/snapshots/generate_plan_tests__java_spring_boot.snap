---
source: tests/generate_plan_tests.rs
expression: plan
---
{
  "nixpacksVersion": "[version]",
  "buildImage": "[build_image]",
  "variables": {
    "NIXPACKS_METADATA": "Java"
  },
  "phases": [
    {
      "name": "install",
      "dependsOn": [
        "setup"
      ]
    },
    {
      "name": "setup",
      "nixPackages": [
        {
          "name": "jdk11"
        },
        {
          "name": "gradle_6"
        }
      ]
    },
    {
      "name": "build",
      "dependsOn": [
        "install"
      ],
      "commands": [
        "./gradlew build"
      ],
      "cacheDirectories": [
        ".gradle"
      ]
    }
  ],
  "startPhase": {
    "cmd": "java $JAVA_OPTS -jar -Dserver.port=$PORT build/libs/*.jar"
  }
}
